File: src/test/java/com/github/rizvn/pgmq/ConsumerTest.java
package com.github.rizvn.pgmq;

import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;

import java.sql.Connection;
import java.sql.ResultSet;
import java.sql.SQLException;
import java.sql.Statement;
import java.util.concurrent.CountDownLatch;
import java.util.concurrent.TimeUnit;

import static org.junit.jupiter.api.Assertions.*;

class ConsumerTest extends DbTest {

    private Consumer consumer;
    private Producer producer;
    private static final String QUEUE_NAME = "test_queue";
    private CountDownLatch messageReceivedLatch;
    private Consumer.PgmqMessage receivedMessage;

    @BeforeEach
    void setUp() throws SQLException {
        // Setup queue for this test
        setupQueue(QUEUE_NAME);

        // Initialize latch for synchronization
        messageReceivedLatch = new CountDownLatch(1);
        receivedMessage = null;

        // Create producer
        producer = new Producer(dataSource);

        // Create consumer with message handler
        consumer = new Consumer(QUEUE_NAME, msg -> {
            // Capture received message
            receivedMessage = msg;
            // Signal that message was received
            messageReceivedLatch.countDown();
        }, dataSource);
    }

    @Test
    void testConsumer() throws SQLException, InterruptedException {
        // Create consumer with message handler
        Consumer testConsumer = new Consumer(QUEUE_NAME, msg -> {
            receivedMessage = msg;
            messageReceivedLatch.countDown();
        }, dataSource);

        // Start consumer
        testConsumer.start();

        // Send a test message
        producer.produce(QUEUE_NAME, "{\"content\": \"Hello, Test!\"}", "{}");

        // Wait for message to be received (with timeout)
        boolean messageReceived = messageReceivedLatch.await(10, TimeUnit.SECONDS);

        // Shutdown consumer
        testConsumer.shutdown();

        // Assert that message was received
        assertTrue(messageReceived, "Expected to receive a message within timeout");
        assertNotNull(receivedMessage, "Expected to have received a message");
        assertEquals("{\"content\": \"Hello, Test!\"}", receivedMessage.getMessage(),
                "Expected message content to match");
    }

    @Test
    void testConsumerWithMultipleMessages() throws SQLException, InterruptedException {
        // Create a latch that waits for 3 messages
        CountDownLatch multiMessageLatch = new CountDownLatch(3);
        final int[] messageCount = {0};

        // Create a new consumer with a message handler that counts messages
        Consumer multiConsumer = new Consumer(QUEUE_NAME, msg -> {
            messageCount[0]++;
            multiMessageLatch.countDown();
        }, dataSource);

        // Start consumer
        multiConsumer.start();

        // Send multiple test messages
        producer.produce(QUEUE_NAME, "{\"msg\": \"Message 1\"}", "{}");
        producer.produce(QUEUE_NAME, "{\"msg\": \"Message 2\"}", "{}");
        producer.produce(QUEUE_NAME, "{\"msg\": \"Message 3\"}", "{}");

        // Wait for all messages to be received
        boolean allMessagesReceived = multiMessageLatch.await(15, TimeUnit.SECONDS);

        // Shutdown consumer
        multiConsumer.shutdown();

        // Assert that all messages were received
        assertTrue(allMessagesReceived, "Expected to receive all 3 messages within timeout");
        assertEquals(3, messageCount[0], "Expected to have processed 3 messages");
    }

    @Test
    void testConsumerArchivesMessages() throws SQLException, InterruptedException {
        // Configure consumer to archive messages after handling
        consumer.archiveAfterHandle = true;

        // Create latch for message received
        CountDownLatch latch = new CountDownLatch(1);
        Consumer archiveConsumer = new Consumer(QUEUE_NAME, msg -> {
            receivedMessage = msg;
            latch.countDown();
        }, dataSource);
        archiveConsumer.archiveAfterHandle = true;

        // Start consumer
        archiveConsumer.start();

        // Send a test message
        producer.produce(QUEUE_NAME, "{\"content\": \"Archive Test\"}", "{}");

        // Wait for message to be received
        boolean messageReceived = latch.await(10, TimeUnit.SECONDS);

        // Shutdown consumer
        archiveConsumer.shutdown();

        // Assert that message was received
        assertTrue(messageReceived, "Expected to receive a message");

        // Verify that the message was archived (not in main queue)
        int mainQueueCount;
        int archiveQueueCount;
        try (Connection conn = dataSource.getConnection();
             Statement stmt = conn.createStatement();
             ResultSet rs = stmt.executeQuery("SELECT count(*) FROM pgmq.q_" + QUEUE_NAME)) {
            assertTrue(rs.next(), "Result set should contain data");
            mainQueueCount = rs.getInt(1);
        }

        try (Connection conn = dataSource.getConnection();
             Statement stmt = conn.createStatement();
             ResultSet rs = stmt.executeQuery("SELECT count(*) FROM pgmq.a_" + QUEUE_NAME)) {
            assertTrue(rs.next(), "Result set should contain data");
            archiveQueueCount = rs.getInt(1);
        }

        // After archiving, the main queue should be empty and archive should have the message
        assertEquals(0, mainQueueCount, "Main queue should be empty after archiving");
        assertEquals(1, archiveQueueCount, "Archive queue should contain the archived message");
    }
}

File: pgmq/Consumer_test.go
package pgmq_test

import (
	"context"
	"database/sql"
	"testing"
	"time"

	"github.com/rizvn/pgutil/pgmq"
	"github.com/rizvn/pgutil/testutil"
)

func TestConsumer(t *testing.T) {
	// create context with timeout
	ctx, cancel := context.WithTimeout(context.Background(), time.Minute*1)

	// Start Postgres test container
	ctr, dsn, err := testutil.StartPgTestContainer()
	if err != nil {
		t.Fatal(err)
	}
	defer func() { _ = ctr.Terminate(context.Background()) }()

	// Create db pool
	dbPool, err := sql.Open("pgx", dsn)
	if err != nil {
		t.Fatalf("failed to create db p pool pool %v", err)
	}
	dbPool.SetMaxOpenConns(20)

	// create consumer
	var recvd *pgmq.PgmqMessage = nil

	c, err := pgmq.NewConsumer(dbPool, "test_queue",
		// message handler runs when a message is received
		func(ctx context.Context, msg *pgmq.PgmqMessage) {
			// capture received message
			recvd = msg

			// cancel context to end test
			cancel()
		})

	if err != nil {
		t.Fatal(err)
	}

	c.Start()

	// create producer
	p := pgmq.NewProducer(dbPool)

	// send a test message
	p.Produce("test_queue", `{"content": "Hello, Test!"}`, "{}")

	// wait for message or context timeout
	for {
		select {
		case <-ctx.Done():
			// check if message was received on cancel or timeout
			if recvd == nil {
				t.Errorf("Expected to receive a message, but none was received")
			}
			// shutdown consumer
			c.ShutdownWithWait()
			return
		}
	}
}
